### PR TITLE
Add support for IPv6 networks (on Linux clients)

### LINKED ISSUE #1167: Feature - Netbird Cloud IPv6
**Is your feature request related to a problem? Please describe.**
We have IPv4 and v6 in our environment and we noticed that v6 is not available as part of the Netbird cloud.

**Describe the solution you'd like**
Please add IPv6 support to Netbird cloud.

**Describe alternatives you've considered**
I could do v4 over v6 tunneling, but thats big complicated and not worth the effort.

**Additional context**

### PR DESCRIPTION
**Pull request by pulsastrix**
## Describe your changes
This PR adds some initial support for assigning IPv6 addresses to hosts inside of the wireguard network.
For now, only Linux using the kernel module and nftables as the firewall implementation is supported.

**Note for reviewers:**
I have been working on this on and off for a couple of months and had to merge in a lot of upstream changes, it's probably easier to review the final version instead of going through each commit separately. :sweat_smile: 

### Notes on address assignment
On network creation (or assignment of the first IPv6 address for existing networks), a random /64 prefix in the ULA range is generated, which will be used for devices on the network.
On enabling IPv6 on supported hosts, a random address will be generated for the host.
Disabling and then re-enabling IPv6 results in a new IPv6 address being generated.

### Testing
I have tested this PR on the following clients:
- Linux (both using kernel wireguard, which supports IPv6, and tun, which does not) 
- Windows (which does not support IPv6)
- Version 0.25.2 on linux combined with the management server built from this branch

Unfortunately, I don't have the environment set up and/or lack the devices to test other operating systems. 

How to test:
1. Spin up an instance of the management server from this branch, as well as an instance of the frontend with the necessary changes (see the corresponding [PR in the dashboard repository](https://github.com/netbirdio/dashboard/pull/314)).
2. Build and run the netbird client from this branch, add it to the management server.
3. Open the frontend, and enable ipv6 in the peer settings.
grafik
4. Alternatively, leave the peer IPv6 setting at "Inherit from Groups" (the default) and instead enable IPv6 for all hosts in a group in the group settings.
grafik
5. Look at the Wireguard interface/firewall/route configuration on the clients, test pinging the devices using IPv6, etc.

### Known issues and TODOs

- [x] Update dashboard changes to work with new dashboard versions
- [x] Integrate with Groups (allow enabling IPv6 for an entire Group at once)
- [x] During development, i have seen problems when updating a peer from a non-IPv6-supported version to an IPv6-supported version (other peers do not establish connections properly until they are restarted) 
    - ~~unsure if this is still the case or if other changes i made fixed that by now~~ haven't seen this problem during my testing anymore
- [x] When the nftables-backend is used but IPv6 is disabled by the management server, we still create an (empty) firewall table
    - maybe problematic if the peer does not support IPv6 at all (disabled kernel support, etc.)
- [x] Improve handling of IPv6 routes
    - [x] Prevent creation of IPv6 routes with routing peers ~~or groups~~ that have IPv6 disabled
        - Disabling for groups might be too inflexible, you might actually want to use an existing group with IPv6 disabled for an IPv6 route, if some of the hosts have IPv6 enabled through other means.
    - [x] ~~Disallow disabling IPv6 for peer if it is acting as a routing peer for an IPv6 route~~
        - I instead opted for the approach listed in the next bullet point
    - [x] Automatically disable all IPv6 routes if IPv6 is disabled for a peer.
        - [x] The "Inherit from Groups" setting should be renamed to "Auto" and also automatically enable IPv6 if peer is IPv6 routing peer.
    - [x] Disable routing peers/groups if they no longer support IPv6 (e.g. because of a downgrade)
    - [x] Ensure that only IPv6-enabled group members are distributed as routing peers if group is used as routing peer
        - (unsure if this is already the case)
- [x] Improve handling of IPv6 Nameservers
    - [x] Distribute IPv6 Nameserver IPs to IPv6 enabled peers only
- [x] Fix compilation issues on non-linux targets caused by my last changes
- [x] Write some more test cases, make sure the existing ones all pass
    - [x] Fix existing test cases
    - [x] Write test cases for client
        - Note that due to the [lack of IPv6 support on GitHub Actions runners](https://github.com/actions/runner-images/issues/668#issuecomment-624080758), the IPv6 tests will be skipped there. Make sure to run them locally in order to verify that they pass.
    - [x] Write test cases for management server
- [x] Documentation
    - can be found in netbirdio/docs#165

Adding support for other operating systems/Linux with iptables/uspfilter is probably something better suited for a separate PR, considering the current size of this one.
I have done some preliminary testing with enabling IPv6 on Windows in one of the older commits - assignment of IPv6 addresses is pretty trivial. However, due to some issue (probably routing), I wasn't able to ping other peers, which is why I did not include it for this PR.

## Issue ticket number and link
Resolves (partially) #1167 and #46 

### Checklist
- [ ] ~~Is it a bug fix~~
- [ ] ~~Is a typo/documentation fix~~
- [x] Is a feature enhancement
- [ ] ~~It is a refactor~~
- [x] Created tests that fail without the change (if possible)
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

**Comment by mlsmaycon** [comment]
Hello @pulsastrix, thanks for the fantastic contribution! 

Would it be possible for you to join our [Slack workspace](https://join.slack.com/t/netbirdio/shared_invite/zt-vrahf41g-ik1v7fV8du6t0RwxSrJ96A) to discuss a few points for this feature? Maybe we can schedule a short call as well.

**Comment by pulsastrix** [comment]
Quick status update:
The current state of this PR now includes all changes from the main branch up until February 25th.

Additionally, the "Assign IPv6 address by default" setting has now been replaced by a groupwide IPv6 toggle, i.e. it is now possible to enable IPv6 for an entire group at once.
By default, peers will now enable IPv6 if they are a member of at least one group that has IPv6 enabled, which allows enabling IPv6 for all hosts at once by simply enabling this toggle for the "All" group.
If IPv6 is explicitly enabled or disabled for a specific peer, the peer settings have precedence.

Finally, the Dashboard-PR has been rewritten based on the new UI.

I plan on resolving the remaining TODOs and getting this PR in a reviewable state sometime the next week, although i can't make any guarantees.

File: management/server/activity/codes.go
Comment (mlsmaycon) [inline management/server/activity/codes.go]: Please add the codes at the end of the const block

File: management/server/network.go
Comment (mlsmaycon) [inline management/server/network.go:92]: This seems to generate the same range on every execution.

Can you use the Read function from the crypto/rand package?

Also, can you use the range `fd00:b14d:RAND_BYTES::/64` ? Where `b14d` = `bird`

Comment (mlsmaycon) [reply management/server/network.go:92]: This will make the ranges easy to identify and possibly memorize (not that most people wants that)

**Review by mlsmaycon — COMMENTED** [review]
Thanks @pulsastrix for the changes. We will start reviewing the PR. 

I left couple of notes already.

File: management/server/activity/codes.go
Comment (mlsmaycon) [reply management/server/activity/codes.go]: We still using `iota` and that would mess with the audit events in place.

File: management/server/network.go
Comment (pulsastrix) [reply management/server/network.go:92]: > This seems to generate the same range on every execution.
> 
> Can you use the Read function from the crypto/rand package?

I have replaced the function call-specific RNG for all IP address generation functions with a global one that is seeded based on `"crypto/rand".Read()`.
With this change, the IP addresses and network prefixes are now different on every call.

> Also, can you use the range `fd00:b14d:RAND_BYTES::/64` ? Where `b14d` = `bird`

Done.

Comment (mlsmaycon) [reply management/server/network.go:92]: thanks

**Comment by pulsastrix** [comment]
Aside from the documentation (which I will probably write tomorrow) this pull request should now be ready for review.

Question for the maintainers: As far as I can tell the methods `DefaultAccountManager.GroupAddPeer()` and `DefaultAccountManager.GroupDeletePeer()` seem to be dead code. Is that intentional?

**Comment by pulsastrix** [comment]
Documentation has now been written and can be found in netbirdio/docs#165.

**Comment by scriptjs** [comment]
@pulsastrix, will these changes support [routing traffic to private ipv6 networks](https://docs.netbird.io/how-to/routing-traffic-to-private-networks)?

**Comment by dpy013** [comment]
hi @pulsastrix 
 Recommended fixes before merging
This branch has conflicts that must be resolved
This error message

**Comment by lixmal** [comment]
If you need help/guidance to merge the new routemanager changes, let me know

**Comment by scriptjs** [comment]
@pulsastrix, @mlsmaycon What sort of time line are we looking at to see this merged?

**Comment by mlsmaycon** [comment]
@@scriptjs, we are looking into focusing on this feature starting next month. We will take over the amazing work done by @pulsastrix and add IPv6 to further clients. 

@pulsastrix, we know you have many conflicts now; let us know if you need help resolving them.

**Comment by pulsastrix** [comment]
Sorry for the delay, I was unfortunately quite busy with other stuff.

I have now merged in all of the upstream changes from the main branch.
Hopefully this merge didn't break anything, the tests still seem to pass.

**Comment by mlsmaycon** [comment]
> Sorry for the delay, I was unfortunately quite busy with other stuff.
> 
> I have now merged in all of the upstream changes from the main branch. Hopefully this merge didn't break anything, the tests still seem to pass.

No need to worry, thanks for going over the conflicts. We will review it ASAP.

**Comment by scriptjs** [comment]
@mlsmaycon Any chance you will get to this soon. Has been a long road to this point and would like to a release with the changes.

**Review by mlsmaycon — APPROVED** [review]
Thanks for the contribution @pulsastrix. 

Due to the amount of diff between the code and our current changes, we will take a different approach for the IPv6 and release the feature when all clients have support to it later this year.

**Comment by scriptjs** [comment]
@mlsmaycon Can you provide an update on this.

**Comment by fruworg** [comment]
Any updates? @mlsmaycon

**Comment by kyberorg** [comment]
Hello,

Like @clauverjat-h  I found Netbird really promising to replace Tailscale or ZeroTier for example. But for me IPv6 is essential.  Additionally, option to use public IPv6 range instead of ULA.

**Comment by arthur-lbchn** [comment]
Hi, IPv6 support is mandatory nowadays, would be happy to have it supported by Netbird

**Pull request merged**
